Allow users to embed data a Request

davidbirdsong opened the pull request:
I need to attach some data to every request, would you consider a small [change to the  Request interface](https://github.com/davidbirdsong/vulcan/commit/3dbdf89308f852f30e319e50d119a280a963cb74)?

davidbirdsong commented:
@klizhentas any thoughts on this? changing interface has a far reaching effect. i'd can submit the PR to implement the extra method where needed.

klizhentas commented:
Can you change it to `GetUserData() interface{}` and `SetUserData(interface{})` otherwise looks good

klizhentas commented:
As a second thought, why not create a special implementation of the interface that has the data you need? Can you elaborate on the scenario when you need this?

davidbirdsong commented:
I'd like to store arbitrary structs that would be too specific for the project as a whole, but I could see others wanting to attach data to requests as they pass through various stages without having to store in the http.Request object.

MDrollette commented:
This seems relevant http://www.gorillatoolkit.org/pkg/context

klizhentas commented:
@davidbirdsong I think that makes sense, so `GetUserData() interface{}` and `SetUserData(interface{})` would work then.

pquerna commented:
People end up needing multiple keys and you can have a chain of middlewares/handlers all using the same request object.... so you end up with `GetUserData(key string) interface{}` and `SetUserData(key string, baton interface{})`

klizhentas commented:
@pquerna makes sense

davidbirdsong commented:
@pquerna so is that the interface? Require a key for setting data and one for getting? I suppose the implementation underneath could handle it's own mapping?

If I understand it correctly, I'll send a PR.

klizhentas commented:
@davidbirdsong I think this would work fine. Setter and Getter with a Key with a map behind the scenes

davidbirdsong commented:
So I was about to go off and write this and then I realized that storing a key on my Middleware defeats the purpose of burying a UserData object on the request. The request is the unit of work and is basically what allows a very stateless approach the logic I'd like to add. If I store a key, I could just store the user data too.

Vulcan already handles when to expire out data, I presume at least, it know better than my Middleware code when a request object is done and dereferences it sufficiently to let the GC grab it at some point.

How would a Middleware know when a key is no longer needed and ready to be deleted? This was why I wanted to dump a data structure on the request, so that my Middleware doesn't have to clean up stale, orphaned data.

Let me know if this makes sense.

klizhentas commented:
Hey David,

I think what we are suggesting is to add the following methods to the request interface:

https://github.com/mailgun/vulcan/blob/master/request/request.go#L13

```
GetUserData(key string) (interface{}, bool)
SetUserData(key string, baton interface{})
DeleteUserData(key string)
```

with underlying implementation:

https://github.com/mailgun/vulcan/blob/master/request/request.go#L53

```
type BaseRequest struct {
    HttpRequest *http.Request
    Id          int64
    Body        netutils.MultiReader
    Attempts    []Attempt
    userDataMutex *mutex.Mutex
    userData map[string]interface{}
}
```

Paul suggested to protect it by mutex in the default implementation, I tend to agree.

Does it make sense?

EDIT: Updated the interface to match map and a couple of typos

klizhentas commented:
EDIT: Updated the interface to match map and a couple of typos

davidbirdsong commented:
It makes sense, but it defeats the usefulness. A concrete example might help to explain why.

Let's say that my middlware implements both `ProcessRequest` and `ProcessResponse`.  The job of the middleware is to use information in the inbound request to completely rewrite the http.Request object. When it completes, all the values are replaced with new values.  In `ProcessResponse` I want to intercept and tee off response body and header data. The original information on the request is now interesting to me again and I'd want to include it in my custom processing. It would be nice to have a scratch space to store some of that original information on the Request object before it gets lost in `ProcessRequest`.

Since a request implements `GetId()`, I could use this value to key into a mapping in my middleware to store to and retrieve this original info. That would totally work and I would have been content with that until I noticed that other phases like `Loadbalancer` and `Endpoint` started wanting access to the original info.  To grant them access to the data, my middleware internals started leaking into my implementation of `Loadbalancer` and `Endpoint`. I figure this original info belongs on the `Request` object, but requiring a key defeats that usefulness.

@pquerna mentioned needing keys in the context of Middleware chaining which I'm slightly ignorant of, so perhaps it makes more sense there.

Put another way. what key would best identify the request data a caller would be interested in retreiving: `Request.GetID()`?...that's sorta conter intuitive.

Sorry this is dragging on, I've built some cool stuff around Vulcan.

klizhentas commented:
David,

Ritght, so with your example that would work like this:

Before rewriting the request in ``

```
func (m *MyMiddleware) ProcessRequest(r Request) {

    // Preserve the original request before rewriting it
    r.SetUserData("originalRequest", copyRequest(r))
}
```

```
func (m *MyMiddleware) ProcessResponse(r Request) {

    // Now I can access the original request if I need it
   originalRequest, exists := r.GetUserData("originalRequest")
   if ! exists {
        // err handling
    }
    // now you have access to your original request

}
```

So basically with this approach your request is the context that allows to attach arbitrary data to the request by any key. 

No need to delete the data as it will go away once request will be GCed.

Am I missing something?

> Sorry this is dragging on, I've built some cool stuff around Vulcan.

No worries at all, I'm eager to work on a better solution that suits your needs!

davidbirdsong commented:
Yep, I follow that.

I now have to track: `"originalRequest"` somewhere and map it to a request ID. If any other phases want to get at this data they'd do the same thing. Then that map would have to be cleaned out as requests completed.

This is what I'd probably do:

``` golang
   originalRequest, exists := r.GetUserData(strconv.FormatInt(r.GetId()))
```

...it works and puts all state in the Request object like I'd prefer, but it's pretty odd--don't you think?

klizhentas commented:
David,

But the request instance is not shared between request, so the `Request` is a different object for every request. 

That means that it's safe to set "originalRequest" key as it would be a different value in different requests.

davidbirdsong commented:
I get it now.

So the key could be static because the mapping is stored on each request. I was thinking "orignalRequest" needed to be unique, but it could be the name of my middleware or something generic and re-used on every request?

klizhentas commented:
Yep, exactly

davidbirdsong committed:
provide user data storage on request interface

expand Request interface to provide 3 new methods:

- SetUserData
- GetUserData
- DeleteUserData

intended to allow middleware to store/fetch/delete arbitrary data on the request, but
other phases are free to do so as well

klizhentas commented on request/request.go:
```diff
@@ -92,3 +108,22 @@ func (br *BaseRequest) GetLastAttempt() Attempt {
+func (br *BaseRequest) GetUserData(key string) (i interface{}, b bool) {
+	br.userDataMutex.RLock()
+	defer br.userDataMutex.RUnlock()
+	i, b = br.userData[key]
```
This operation will crash if you haven't executed `SetUserData` before executing `GetUserData` as `userData` will bi nil, right?

davidbirdsong commented:
Good catch--just pushed up a change.

I'm opting for not calling `make` unless the user wants to actually store something since a Request object has a high rate of creation.

klizhentas commented:
Thanks! One last question - can you please add a specific test case that covers Get without prior Set so we have this behavior enforced in test.

Otherwise looks good!

davidbirdsong commented:
Yep, good call. I've added that test.
